# 1. Import Libraries

In [23]:

import numpy as np
import pandas as pd
import os
import kagglehub

# 2. Introduction

Retail sales data contains valuable information about customer purchasing behavior, product performance, sales trends, profitability, and business operations. Analyzing this data helps organizations understand customer preferences, identify high-performing products, optimize inventory management and make data-driven business decisions.

In this project, a retail sales dataset containing **4,310 records and 21 attributes** was analyzed using Python libraries such as **Pandas**, **NumPy**, **Matplotlib** and **Seaborn**. The primary objective of this analysis is to clean and preprocess the raw data, perform exploratory data analysis (EDA), identify meaningful patterns and generate visual insights that support business decision-making.

Before performing any analysis, the dataset was thoroughly examined to identify data quality issues such as missing values, duplicate records, inconsistent formatting, invalid values, and incorrect data types. These issues were resolved through systematic data cleaning and preprocessing to ensure the dataset was accurate, consistent and suitable for further analysis and visualization.

---

# 3. Dataset Overview

The retail sales dataset contains transactional information related to customer purchases and sales activities. Each record represents a unique retail transaction and includes customer information, product details, pricing, sales performance, shipping information, payment methods and customer feedback.

## 3.1 Dataset Summary

- **Number of Records (Rows):** 4,310
- **Number of Attributes (Columns):** 21
- **Dataset Type:** Retail Sales Transactions
- **File Format:** CSV
- **Analysis Tools:** Python, Pandas, NumPy, Matplotlib, Seaborn

## 3.2 Dataset Attributes

| Column Name | Description |
|-------------|-------------|
| `order_id` | Unique identifier for each order |
| `order_date` | Date on which the order was placed |
| `customer_id` | Unique identifier of the customer |
| `customer_name` | Name of the customer |
| `age` | Customer's age |
| `gender` | Gender of the customer |
| `region` | Customer's region |
| `city` | Customer's city |
| `product_category` | Category of the purchased product |
| `product_name` | Name of the purchased product |
| `quantity` | Number of items purchased |
| `unit_price` | Price of one unit of the product |
| `discount_pct` | Discount percentage applied |
| `sales_amount` | Total sales amount |
| `profit` | Profit earned from the transaction |
| `shipping_cost` | Shipping charges |
| `payment_method` | Mode of payment |
| `customer_satisfaction` | Customer satisfaction rating |
| `return_flag` | Indicates whether the product was returned |
| `order_status` | Current order status |
| `days_to_ship` | Number of days taken for product delivery |


## 3.3 Objectives of the Analysis

The primary objectives of this project are:

- To understand the overall structure and quality of the retail sales dataset.
- To identify and resolve missing values, duplicate records, invalid entries, and inconsistent data formats.
- To standardize categorical variables and convert columns to appropriate data types.
- To prepare the dataset for Exploratory Data Analysis (EDA).
- To analyze sales performance, customer purchasing behavior, product trends, and profitability.
- To create meaningful visualizations for better business understanding.
- To derive actionable business insights that can support informed decision-making.

# 4. Load Dataset

In [24]:
df=pd.read_csv("/kaggle/input/datasets/satyakidas07/retail-sales-dataset/retail_sales_dataset.csv")

# 5. Initial Data Exploration

In [25]:
df.head()

,order_id,order_date,customer_id,customer_name,age,gender,region,city,product_category,product_name,...,unit_price,discount_pct,sales_amount,profit,shipping_cost,payment_method,customer_satisfaction,return_flag,order_status,days_to_ship
0,ORD-03903,2024-08-15,CUST32603,Aarav Mehta,39.0,Other,Central,Raipur,Groceries,Sugar,...,346.32,NaN,4043.14,289.18,56.04,Debit Card,NaN,True,Returned,7.0
1,ORD-02198,2022-08-16,CUST07881,Sanjay Verma,28.0,Male,West,Mumbai,Beauty,Sunscreen,...,858.66,0.08,9147.31,5355.56,64.49,Debit Card,5.0,False,Delivered,2.0
2,ORD-01348,2021-08-29,CUST49296,Pooja Sharma,41.0,Other,West,Surat,Furniture,Sofa,...,25941.32,0.14,110156.97,36956.51,67.65,Debit Card,4.0,False,Delivered,5.0
3,ORD-02076,2022-07-05,CUST05587,Priya Reddy,29.0,Other,West,Ahmedabad,Furniture,Bed Frame,...,22071.92,0.17,262066.11,66644.99,54.47,EMI,3.0,False,Delivered,4.0
4,ORD-00287,2020-05-07,CUST58035,Ritu Sharma,9.0,Male,Central,Nagpur,Groceries,Cooking Oil,...,320.06,0.31,323.95,31.38,48.81,Net Banking,4.0,False,Pending,4.0


In [26]:
print("DatasetShape:")
print(df.shape)

DatasetShape:
(4310, 21)


In [27]:
print("DatasetInfo")
print(df.info())

DatasetInfo
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4310 entries, 0 to 4309
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   order_id               4280 non-null   object 
 1   order_date             4280 non-null   object 
 2   customer_id            4280 non-null   object 
 3   customer_name          4280 non-null   object 
 4   age                    4150 non-null   float64
 5   gender                 4280 non-null   object 
 6   region                 4280 non-null   object 
 7   city                   4280 non-null   object 
 8   product_category       4280 non-null   object 
 9   product_name           4280 non-null   object 
 10  quantity               4170 non-null   float64
 11  unit_price             4280 non-null   float64
 12  discount_pct           4143 non-null   float64
 13  sales_amount           4280 non-null   float64
 14  profit                 4280 non-null   float

In [28]:
print("StatisticalSummary")
print(df.describe(include="all"))

StatisticalSummary
         order_id  order_date customer_id customer_name          age gender  \
count        4280        4280        4280          4280  4150.000000   4280   
unique       4200        1837        4112           400          NaN     11   
top     ORD-01253  2021-02-04   CUST00213   Meera Kumar          NaN  Other   
freq            2           9           3            21          NaN   1442   
mean          NaN         NaN         NaN           NaN    35.335181    NaN   
std           NaN         NaN         NaN           NaN    35.709382    NaN   
min           NaN         NaN         NaN           NaN    -7.000000    NaN   
25%           NaN         NaN         NaN           NaN    26.000000    NaN   
50%           NaN         NaN         NaN           NaN    34.000000    NaN   
75%           NaN         NaN         NaN           NaN    42.000000    NaN   
max           NaN         NaN         NaN           NaN   999.000000    NaN   

       region      city product_

In [29]:
print(df.isnull().sum())

order_id                  30
order_date                30
customer_id               30
customer_name             30
age                      160
gender                    30
region                    30
city                      30
product_category          30
product_name              30
quantity                 140
unit_price                30
discount_pct             167
sales_amount              30
profit                    30
shipping_cost             30
payment_method            30
customer_satisfaction    378
return_flag               30
order_status              30
days_to_ship             130
dtype: int64


In [30]:
print(df.duplicated().sum())

109



# 6. Data Quality Summary (Pre-Cleaning)

| Issue | Details |
|--------|---------|
| **Total Rows** | 4,310 |
| **Duplicate Rows** | 109 duplicate records identified and removed using `drop_duplicates()` |
| **Columns with Missing Values** | All 21 columns contained at least some missing (`NaN`) values |
| **Largest Missing-Value Columns** | `customer_satisfaction` (~378 missing values), `quantity` (~167 missing values) |
| **Data Type Mismatches** | `order_id`, `customer_id`, `customer_satisfaction`, and `days_to_ship` were stored as `object` or `float` instead of integer |
| **Prefix Noise** | `order_id` contained the prefix **"ORD-"** and `customer_id` contained the prefix **"CUST"**, which were removed |
| **Date Format Inconsistency** | `order_date` contained mixed date formats and was standardized to the `datetime` format |
| **Invalid Values** | `age` contained values less than 0 and greater than 100. `quantity`, `shipping_cost`, `unit_price`, `sales_amount`, `discount_pct`, `customer_satisfaction`, and `days_to_ship` contained invalid negative values, which were corrected. |
| **Text Inconsistency** | Categorical columns such as `gender`, `city`, `region`, `product_category`, `product_name`, `payment_method`, `return_flag`, and `order_status` contained inconsistent capitalization and extra spaces, which were standardized. |

---

# 7. Data Cleaning & Preprocessing

## 7.1 Remove Duplicate Records

In [31]:
df.drop_duplicates(inplace=True)

## 7.2 Handle Invalid Values

In [32]:
df.loc[(df["age"]<0)|(df["age"]>100),"age"]=np.nan

non_negative_c = [
    "quantity",
    "unit_price",
    "discount_pct",
    "sales_amount",
    "shipping_cost",
    "days_to_ship",
    "customer_satisfaction"
]

for col in non_negative_c:
    df.loc[df[col] < 0, col] = np.nan

## 7.3 Handle Missing Values

In [33]:

df["age"]=df["age"].fillna(df["age"].mean()).astype(int)
numeric_columns=[
"unit_price","sales_amount","profit","shipping_cost","quantity","discount_pct","customer_satisfaction","days_to_ship"
]
for col in numeric_columns:
    df[col]=df[col].fillna(df[col].median())
print(df.describe())

               age     quantity    unit_price  discount_pct  sales_amount  \
count  4201.000000  4201.000000   4201.000000   4201.000000  4.201000e+03   
mean     34.208760     7.364913  10700.716669      0.202140  6.290870e+04   
std      11.674575    43.414984  17407.192855      0.116113  3.194756e+05   
min       0.000000     0.000000     50.260000      0.000000  3.945000e+01   
25%      26.000000     3.000000    680.850000      0.100000  2.473000e+03   
50%      34.000000     5.000000   2246.005000      0.210000  9.647895e+03   
75%      42.000000     8.000000  12824.110000      0.300000  5.050240e+04   
max      72.000000   999.000000  79957.920000      0.400000  1.736031e+07   

             profit  shipping_cost  customer_satisfaction  days_to_ship  
count  4.201000e+03    4201.000000            4201.000000   4201.000000  
mean   1.178822e+04      63.410787               3.016425      5.639848  
std    5.334600e+04      18.256726               1.354346      4.141718  
min    1.5

## 7.4 Clean Identifier Columns

In [34]:
# Remove prefixes
df["order_id"] = (
    df["order_id"]
    .astype(str)
    .str.replace("ORD-", "", regex=False)
)

df["customer_id"] = (
    df["customer_id"]
    .astype(str)
    .str.replace("CUST", "", regex=False)
)

# Convert to numeric (invalid values become NaN)
df["order_id"] = pd.to_numeric(df["order_id"], errors="coerce")
df["customer_id"] = pd.to_numeric(df["customer_id"], errors="coerce")

# Drop rows where either ID is missing
df.dropna(subset=["order_id", "customer_id"], inplace=True)

## 7.5 Standardize Text Columns

In [35]:

object_columns = ["customer_name","gender","region","city","product_category","product_name","payment_method","return_flag","order_status"]

for col in object_columns:
    df[col]=(df[col].astype(str).str.strip().str.title())


In [36]:
df["gender"]=df["gender"].replace({
"male":"Male",
"MALE":"Male",
"m":"Male",
"M":"Male",
"female":"Female",
"FEMALE":"Female",
"f":"Female",
"F":"Female",
"other":"Other",
"OTHER":"Other"
})
print(df["gender"].value_counts())

gender
Other     1419
Male      1408
Female    1373
Name: count, dtype: int64


## 7.6 Convert Data Types

In [44]:
df["order_id"] = df["order_id"].astype(int)
df["customer_id"] = df["customer_id"].astype(int)
df["customer_satisfaction"] = pd.to_numeric(df["customer_satisfaction"], errors="coerce").astype(int)
df["days_to_ship"] = pd.to_numeric(df["days_to_ship"], errors="coerce").astype(int)
df["quantity"] = pd.to_numeric(df["quantity"], errors="coerce").astype(int)
df["order_date"] = pd.to_datetime(df["order_date"],format="mixed",dayfirst=True,errors="coerce")

## 7.7 Final Validation

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4200 entries, 0 to 4309
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   order_id               4200 non-null   int64         
 1   order_date             4200 non-null   datetime64[ns]
 2   customer_id            4200 non-null   int64         
 3   customer_name          4200 non-null   object        
 4   age                    4200 non-null   int64         
 5   gender                 4200 non-null   object        
 6   region                 4200 non-null   object        
 7   city                   4200 non-null   object        
 8   product_category       4200 non-null   object        
 9   product_name           4200 non-null   object        
 10  quantity               4200 non-null   int64         
 11  unit_price             4200 non-null   float64       
 12  discount_pct           4200 non-null   float64       
 13  sales_am

In [39]:
df.describe()

,order_id,order_date,customer_id,age,quantity,unit_price,discount_pct,sales_amount,profit,shipping_cost,customer_satisfaction,days_to_ship
count,4200.000000,4200,4200.000000,4200.000000,4200.000000,4200.000000,4200.000000,4.200000e+03,4.200000e+03,4200.000000,4200.000000,4200.000000
mean,2100.500000,2022-07-06 12:29:08.571428608,49942.747857,34.208810,7.365476,10702.729695,0.202138,6.292138e+04,1.179028e+04,63.410774,3.016429,5.639762
min,1.000000,2020-01-01 00:00:00,13.000000,0.000000,0.000000,50.260000,0.000000,3.945000e+01,1.560000e+00,0.510000,1.000000,0.000000
25%,1050.750000,2021-04-20 00:00:00,25059.750000,26.000000,3.000000,680.817500,0.100000,2.472295e+03,7.120700e+02,50.855000,2.000000,3.000000
50%,2100.500000,2022-07-13 00:00:00,49522.000000,34.000000,5.000000,2246.005000,0.210000,9.647895e+03,3.169875e+03,63.465000,3.000000,6.000000
75%,3150.250000,2023-09-27 00:00:00,74380.000000,42.000000,8.000000,12828.977500,0.300000,5.052485e+04,1.051096e+04,75.655000,4.000000,8.000000
max,4200.000000,2024-12-30 00:00:00,99976.000000,72.000000,999.000000,79957.920000,0.400000,1.736031e+07,2.445648e+06,133.350000,5.000000,100.000000
std,1212.579894,NaN,28871.602575,11.675964,43.420138,17408.776463,0.116126,3.195126e+05,5.335219e+04,18.258900,1.354508,4.142208


In [40]:
df.isnull().sum()

order_id                 0
order_date               0
customer_id              0
customer_name            0
age                      0
gender                   0
region                   0
city                     0
product_category         0
product_name             0
quantity                 0
unit_price               0
discount_pct             0
sales_amount             0
profit                   0
shipping_cost            0
payment_method           0
customer_satisfaction    0
return_flag              0
order_status             0
days_to_ship             0
dtype: int64

In [41]:
df.duplicated().sum()

np.int64(0)

## 7.8  Save Clean Dataset

In [43]:
df.to_csv("cleaned_retail_sales_dataset.csv", index=False)